### PR TITLE
feat!: Require schemas and object-style rules

### PR DESCRIPTION
**Pull request by bmish**
## Summary

We propose two changes to rules:

- Requiring rules with options to specify schemas (removing support for rules with options that are missing schemas)
- Requiring rule implementations to use the object-style format (removing support for the deprecated function-style format)

## Demonstrations

- Deprecation warnings when running rule tests: https://github.com/bmish/eslint/tree/RFC-85-deprecation-warning

## Related Issues

* https://github.com/eslint/eslint/issues/14709

### CODE REVIEW

File: designs/2021-schema-object-rules/README.md
Comment (nzakas) [inline designs/2021-schema-object-rules/README.md:69]: Note that these changes will need to be made in the flat config files as we aren’t making any further changes to the eslintrc format.

File: designs/2021-schema-object-rules/README.md
Comment (nzakas) [inline designs/2021-schema-object-rules/README.md:172]: Would updates to RuleTester help?

**Review by nzakas — CHANGES_REQUESTED** [review]
Overall very much in favor of these changes. Just a few notes.

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md:172]: One of the items in this list is:

> Adding deprecation notices inside ESLint in a minor version release, showing a warning when executing linting or running tests for an offending rule

So by that I mean the deprecation notice would show when running rule tests. Does that cover what you are suggesting here?

Comment (bmish) [reply designs/2021-schema-object-rules/README.md:172]: I have added some more details about the deprecation notices when running tests or executing linting for an offending lint rule.

I have also started a branch to demonstrate the deprecation warnings: https://github.com/bmish/eslint/tree/RFC-85-deprecation-warning

File: designs/2021-schema-object-rules/README.md
Comment (aladdin-add) [inline designs/2021-schema-object-rules/README.md:43]: I'm 👍to change the default to ` []`, but do we want to add a new way to allow rule authors explicitly skipping the validating - e.g. `schema: false`?
I'm asking as it could happen in some rare cases, e.g. writing an internal rule.

File: designs/2021-schema-object-rules/README.md
Comment (mdjermanovic) [inline designs/2021-schema-object-rules/README.md:178]: Is this necessary, since we will have deprecation warnings when running tests?

Showing warnings when executing linting could be annoying to end users, as there's nothing they can do to fix the issue.

File: designs/2021-schema-object-rules/README.md
Comment (mdjermanovic) [reply designs/2021-schema-object-rules/README.md:43]: > I'm to change the default to ` []`, but do we want to add a new way to allow rule authors explicitly skipping the validating - e.g. `schema: false`?

A workaround can be `schema: {}` or `schema: { type: "array" }`

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md:178]: I was hoping that showing the deprecation warning to end users would result in them pinging the maintainer / filing an issue on github / or even opening a PR themselves with the fix. This would mainly be necessary for stagnant eslint plugins where the maintainer might not otherwise notice the deprecation warning since they aren't regularly running tests.

Perhaps I can propose a different timeline however:

1. When this RFC is approved, add the deprecation warning when running tests, in the hopes of most affected plugins seeing this.
2. In the final minor version of ESLint 8, add the deprecation warning for end users when running linting to try to catch any last holdouts.
3. In ESLint 9, change the deprecation warnings into fatal assertions.

File: designs/2021-schema-object-rules/README.md
Comment (mdjermanovic) [inline designs/2021-schema-object-rules/README.md:160]: This will be also a breaking change for end users who have invalid configurations for non-offending rules (those that don't expect any options), as they'll have to remove useless options from config files.

File: designs/2021-schema-object-rules/README.md
Comment (mdjermanovic) [reply designs/2021-schema-object-rules/README.md:178]: I'd vote for not showing warnings to end users since we have other ways to notify maintainers: deprecation warnings when running tests, blog posts, and prerelease versions. If all of that fails to notify some plugins, the worst case is that they'll be notified by users when we release ESLint 9 final.

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md:43]: I could see both sides, but my current preference is to *not* provide an explicit opt-out.

Why we shouldn't give users an explicit schema opt-out:
- An easy, well-known option to opt-out is just welcoming laziness, and many plugins authors may default to always using it, despite the benefits and ease of writing a real schema, defeating many of the benefits described in this RFC
- An explicit opt-out option is essentially encouraging/recommending opting-out as an acceptable approach
- As @mdjermanovic suggested, users can use the the workaround of providing an incomplete schema like`schema: {}` or `schema: { type: "array" }` if they really want to opt-out
- In my opinion, schemas should be low-effort and valuable for all rules, even internal rules

Why we should give users an explicit schema opt-out like `schema: false`:
- If users are going to abuse obscure workarounds like `schema: {}` or `schema: { type: "array" }` anyway, then it would be clearer for everyone if they could explicitly specify that they don't want a schema

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md]: I added a bullet point to denote this.

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md]: Agreed with @mdjermanovic. The prerelease versions should give users/plugins time to discover any remaining problems. Updated RFC to only add a deprecation warning when running tests.

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md]: I added a paragraph about this workaround in the `Drawbacks of requiring schemas` section including touching on why I am against an explicit opt-out.

File: designs/2021-schema-object-rules/README.md
Comment (nzakas) [reply designs/2021-schema-object-rules/README.md:177]: I don’t think deprecation noticed targeted at end users are a useful approach. How many will actually hunt down the author of a rule (that was maybe included in a shareable config) to let them know?

It seems like targeted messaging in RuleTester of some sort would reach the correct audience.

Comment (bmish) [reply designs/2021-schema-object-rules/README.md:177]: Yep, addressed as part of the other comment thread: https://github.com/eslint/rfcs/pull/85#discussion_r777130444

Comment (nzakas) [reply designs/2021-schema-object-rules/README.md:177]: Doh sorry, missed that. Just digging out of my holiday notifications!

**Review by nzakas — APPROVED** [review]
I’m happy with the overall direction at this point. 👍

File: designs/2021-schema-object-rules/README.md
Comment (mdjermanovic) [reply designs/2021-schema-object-rules/README.md]: Both arguments for and arguments against `schema: false` are good.

> Why we should give users an explicit schema opt-out like `schema: false`:
> 
> * If users are going to abuse obscure workarounds like `schema: {}` or `schema: { type: "array" }` anyway, then it would be clearer for everyone if they could explicitly specify that they don't want a schema

I suggested the workarounds, but this is a very good argument and I'm now for an explicit opt-out like `schema: false`.

As described in the Motivation section of this RFC, one of the benefits of rule schemas (and thus forcing rules to define schemas) is automated analysis of rule options, like auto-generated docs, auto-complete, maybe even detecting breaking changes as mentioned in https://github.com/eslint/eslint/issues/15428, etc., For that purpose, when some rule doesn't provide a schema for whatever reason, it seems better to see that right away from `schema: false` than to analyze a meaningless workaround schema.

Also, some schema validations we might want to add to RuleTester in the future could report these workaround schemas as potentially incorrect, since they're essentially a no-op.

**Review by mdjermanovic — COMMENTED** [review]
Aside from resolving the discussion about schema opt-outs (https://github.com/eslint/rfcs/pull/85#discussion_r775110002), everything else looks good to me, and the analysis of plugins is awesome!

File: designs/2021-schema-object-rules/README.md
Comment (btmills) [reply designs/2021-schema-object-rules/README.md]: Also in favor of an explicit `schema: false` opt-out. I can imagine some companies' internal rules that aren't accessible to the (excellent, incredibly helpful) ecosystem analyzer don't have schemas because they're only configured once by the person who wrote the rule. Having `schema: false` means `eslint-plugin-eslint-plugin` could trivially add a warning-level rule to encourage finishing the schema rather than having to analyze no-op schemas.

**Review by btmills — APPROVED** [review]
LGTM pending addition of `schema: false`. Thank you for building the ecosystem analyzer; this is the sort of change I would be very nervous about making without data!

**Comment by nzakas** [comment]
@mdjermanovic do you feel comfortable moving to final commenting?

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md]: Sounds good, I switched to use `schema: false`.

**Review by mdjermanovic — APPROVED** [review]
Looks good to me, thanks!

File: designs/2021-schema-object-rules/README.md
Comment (ljharb) [inline designs/2021-schema-object-rules/README.md:14]: This is a good change; a few of my plugins have rules that lack a schema because it’s a breaking change, but my planned workaround for those is to conditionally use an empty schema for eslint 9+

File: designs/2021-schema-object-rules/README.md
Comment (ljharb) [inline designs/2021-schema-object-rules/README.md:30]: I believe `schema: {}` works as well?

Comment (bmish) [reply designs/2021-schema-object-rules/README.md:30]: I just tested to confirm that `schema: {}` does not actually enforce anything. As mentioned in comments like https://github.com/eslint/rfcs/pull/85#discussion_r809932236, we should probably disallow no-op schemas like that. If someone wants to enforce that there are no options, they should omit the schema (after this RFC) or use `schema: []`. If someone wants to skip writing a schema, they will be able to use `schema: false`.

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md]: One follow-up to this, I think we should disallow `schema: {}`. That is the simplest no-op schema and can be easily confused with `schema: []` which actually does enforce that there are no options. Anyone using `schema: {}` is likely making a mistake, so we can throw an error if we detect that:

```
`schema: {}` is a no-op. For rules with options, please fill in a complete schema. For rules without options, please omit `schema` or use `schema: []`.
```

We could also disallow other obvious ones like `schema: { type: "array" }` but I'm not too worried about catching all possible no-op schemas right now.

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md:15]: I just did some testing and found that ESLint v1 only supported function-style rules. It looks like ESLint v2 began supporting object-style rules too.

I'm pretty confident that function-style rules do not offer any extra functionality over object-style rules.

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md]: `schema: { additionalProperties: false }` is also a no-op.

Comment (ljharb) [reply designs/2021-schema-object-rules/README.md]: ah, maybe `{ maxProperties: 0 }` or something, but clearly `[]` is simpler

Comment (bmish) [reply designs/2021-schema-object-rules/README.md]: I updated the `Opt-out` section to mention that we will disallow a few common no-op schemas.

**Review by aladdin-add — APPROVED** [review]
LGTM~

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: designs/2021-schema-object-rules/README.md
Comment (bmish) [reply designs/2021-schema-object-rules/README.md:106]: If it causes a problem, I'm happy to revert this new exception. I figured this RFC would be a good chance to include it since it's very related, and you suggested it originally in the other thread. Definitely want to get the full TSC's opinion on it. I can fill in more details soon.

Comment (nzakas) [reply designs/2021-schema-object-rules/README.md:106]: In general, once an RFC is in final commenting, we shouldn’t be making any further additions, only adding clarifications.

I think it makes sense to bundle this change with the rest of the RFC but the additional details @mdjermanovic mentioned would be helpful.

Comment (bmish) [reply designs/2021-schema-object-rules/README.md:106]: 1. We can throw the exception in both core and rule tester. In core, we can throw it next to `ajv.compile` where the existing schema validation happens.
2. Honestly, I'm fine just checking for `schema: {}` since that's the most-common mistake and most-easy to mix up with `schema: []`. It should be easy to add more to the list later if there's an interest or need.
3. Yes I think we can log a warning in ESLint v8 when running tests for a rule with such a schema.

I used [eslint-ecosystem-analyzer](https://github.com/bmish/eslint-ecosystem-analyzer) and found that about ~7 of the top 1,000 plugins have an occurrence of `schema: {}`.

Comment (aladdin-add) [reply designs/2021-schema-object-rules/README.md:106]: I agreed, there is nothing the plugin users can do to avoid the error.  they could only disable the rule totally and report to the plugin author.

Comment (mdjermanovic) [reply designs/2021-schema-object-rules/README.md]: > Although note that other schema validation happens in core already

Do we have validations of rule schemas in core? I think we've disabled Ajv validations and that we don't have any custom validations.

Comment (bmish) [reply designs/2021-schema-object-rules/README.md]: My bad, I just confirmed that we don't validate schemas in core. We do `ajv.compile` in core but `ajv.validateSchema` only in rule tester.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
